proposals/collection-expression-arguments.md

```diff
@@ -0,0 +1,184 @@
+
+Examinations of the solutions for just passing *comparers* have revealed weaknesses in their approach if we wanted
+ to expand them to *arbitrary arguments*.  For example:
+
+1. Reusing *element* syntax, like we do with the form: `[StringComparer.OrdinalIgnoreCase, "mads": 21]`. This works
+   well in a space where `KeyValuePair<,>` and comparers do not inherit from common types.  But it breaks down in a
+   world where one might do: `HashSet<object> h = [StringComparer.OrdinalIgnoreCase, "v"]`.  Is this passing along
+   a comparer?  Or attempting to put two object values into the set?
+
+2. Separating out arguments versus elements with subtle syntax (like using a semicolon instead of a comma to
+   separate them in `[comparer; v1]`). This risks very confusing situations where a user accidentally writes `[1; 2]`
+   (and gets a collection that passes '1' as, say, the 'capacity' argument for a `List<>`, and only contains the
+   single value '2'), when they intended `[1, 2]` (a collection with two elements).
+
+Because of this, in order to support arbitrary arguments, we believe a more obvious syntax is needed to more
+clearly demarcate these values. Several other design concerns have also come up with in this space.  In no
+particular order, these are:
+
+1. That the solution not be ambiguous and cause breaks with code that people are likely using with collection
+ expressions today.  For example:
+
+    ```c#
+    List<Widget> c = [new(...), w1, w2, w3];
+    ```
+    
+    This is legal today, with the `new(...)` expression being a 'implicit object creation' that creates a new
+    widget.  We cannot repurpose this to pass along arguments to `List<>`'s constructor as it would *certainly*
+    break existing code.
+
+1. That the syntax not extend to outside of the `[...]` construct.  For example:
+
+    ```c#
+    HashSet<string> s = [...] with ...;
+    ```
+    
+    These syntaxes can be construed to mean that the collection is created first, and then recreated into a
+    differing form, implying multiple transformations of the data, and potentially unwanted higher costs
+    (even if that's not what is emitted).
+
+1. That `new` as a potential keyword to use *at all* in this space is undesirably confusing.  Both because
+   `[...]` *already* indicates that a *new* object is created, and because translations of the collection
+   expression may go through non-constructor APIs (for example, the *Create method* pattern).
+
+1. That the solution not be excessively verbose.  A core value proposition of collection expressions is
+   *brevity*. So if the form adds a large amount of syntactic scaffolding, it will feel like a step backwards,
+   and will undercut the value proposition of using collection-expressions, versus calling into the existing
+   APIs to make the collection.
+
+Note that a syntax like `new([...], ...)` runs afoul of both '2' and '3' above.  It makes it appear as if we
```

Comment:
Consider also `new(...) [...]`:
- it doesn't have the disadvantage of `[args(...), ...]` which looks like the `args(...)` is one of the collection elements (since it's inside the `[...]`),
- is consistent with collection initializers `new(...) {...}` ,
- a disadvantage is that it looks like calling a constructor but it can call the builder pattern (then perhaps `args(...) [...]` which is a new syntactic form so it shouldn't be a break)

Reply:
> which is a new syntactic form so it shouldn't be a break)

Technically that's just an invocation that is then indexed into :-)

But we'll consider this. Is just one that we've already looked at and had several things we didn't like about it.